src/Workspaces/Core/Portable/EmbeddedLanguages/DateTime/LanguageServices/DateTimePatternDetector.cs

```diff
@@ -0,0 +1,202 @@
+            var name = GetNameOfInvokedExpression(invokedExpression);
+            if (name != nameof(ToString) && name != nameof(System.DateTime.ParseExact) && name != nameof(System.DateTime.TryParseExact))
+                return false;
+
+            // if we couldn't determine the arg name or arg index, can't proceed.
+            var (argName, argIndex) = GetArgumentNameOrIndex(argumentNode);
+            if (argName == null && argIndex == null)
+                return false;
+
+            // If we had a specified arg name and it isn't 'format', then it's not a DateTime
+            // 'format' param we care about.
+            if (argName != null && argName != FormatName)
+                return false;
+
+            var symbolInfo = _semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken);
+            var method = symbolInfo.Symbol;
+            if (TryAnalyzeInvocation(method, argName, argIndex))
+                return true;
+
+            foreach (var candidate in symbolInfo.CandidateSymbols)
+            {
+                if (TryAnalyzeInvocation(candidate, argName, argIndex))
+                    return true;
+            }
+
+            return false;
+        }
+
+        private (string? name, int? index) GetArgumentNameOrIndex(SyntaxNode argument)
+        {
+            var syntaxFacts = _info.SyntaxFacts;
+
+            var argName = syntaxFacts.GetNameForArgument(argument);
+            if (argName != "")
+                return (argName, null);
+
+            var arguments = syntaxFacts.GetArgumentsOfArgumentList(argument.Parent);
+            var index = arguments.IndexOf(argument);
+            if (index >= 0)
+                return (null, index);
+
+            return default;
+        }
+
+        private bool TryAnalyzeInvocation(ISymbol? symbol, string? argName, int? argIndex)
+            => symbol is IMethodSymbol method &&
+               method.DeclaredAccessibility == Accessibility.Public &&
+               method.MethodKind == MethodKind.Ordinary &&
+               IsDateTimeType(method.ContainingType) &&
+               AnalyzeStringLiteral(method, argName, argIndex);
+
+        private bool IsDateTimeType(INamedTypeSymbol containingType)
+            => _dateTimeType.Equals(containingType) || _dateTimeOffsetType.Equals(containingType);
+
+        private bool AnalyzeStringLiteral(IMethodSymbol method, string? argName, int? argIndex)
+        {
+            Debug.Assert(argName != null || argIndex != null);
+
+            var parameters = method.Parameters;
+            if (argName != null)
+                return parameters.Any(p => p.Name == argName);
+
+            var parameter = argIndex!.Value < parameters.Length ? parameters[argIndex.Value] : null;
+            return parameter?.Name == FormatName;
+        }
+
+        private IParameterSymbol? GetParameter(IMethodSymbol method, SyntaxNode argument)
```

Comment:
I did a cntrl+F on this and didn't find anything - is it used?

Reply:
removed!